### DISCUSSION TITLE
fails to resolve intel-built numpy dependencies

### DISCUSSION THREAD

**Discussion by ecneil** [python-poetry/poetry#7659] [Q&A] [answered]
I tried to install the mkl version of numpy that intel provides at anaconda.org by putting the following in pyproject.toml
```
[tool.poetry.dependencies]
python = "^3.9"
numpy = {version = "1.22.3", source = "intel-build"}

[[tool.poetry.source]]
name = "intel-build"
url = "https://pypi.anaconda.org/intel/simple"
default = false
secondary = false

[[tool.poetry.source]]
name = "intel-build-wheel"
url = "https://pypi.anaconda.org/intel/label/wheels/simple"
default = false
secondary = false
```

When I ran `poetry lock`, it failed to resolve the dependencies as below.

```
Because no versions of mkl-umath match !=0.1.1
 and mkl-umath (0.1.1) depends on numpy (1.19.5), every version of mkl-umath requires numpy (1.19.5).
So, because intelnp-poetry depends on numpy (1.22.3) which depends on mkl_umath (*), version solving failed.
```

After a brief investigation, it appears that the dependency cannot be resolved due to the following reasons.
- When intel updates a package on pypi, if it is a rebuild for the purpose of updating dependencies, only the build tag is updated and re-updated with the same version
- PyPI's JSON api does not seem to reflect information in the requires_dist when the same version of a package with a different build tag is uploaded
- Since poetry uses information from PyPI's JSON api to resolve dependencies, it cannot resolve dependencies if the information is not reflected in the JSON api.

(The PyPI JSON api response says `requires_dist` is `numpy==1.19.5`, but the wheel with build tag 27 has `Requires-Dist: numpy>=1.22.3,<1.23.0` in METADATA)

In the end, should this problem be treated as a problem of poetry, PyPI, or intel?

**Comment by dimbleby** [comment, marked as answer]
all of the above

poetry assumes that all distributions of the same version of a package have the same metadata.  That's not required by the specs, but it is almost always true.  While strictly wrong, that assumption is unlikely to change: because poetry finds cross-platform solutions, dropping it would be _horrible_ for performance (requiring poetry both to download many more wheels / tarballs, and also potentially to solve a much larger problem space).

So poetry is at fault there: but raising an issue asking it to change will not help

pypi's json interface makes the same assumption.  It's wrong too; but it's wrong in a way that's convenient for poetry.

the most likely way to make mkl-umath and poetry compatible would be for mkl-umath to start using regular version numbers instead of everything being labelled 0.1.1, and for them to use consistent metadata across all distributions of those sensibly-versioned packages.

But https://github.com/IntelPython/mkl_umath/commit/667f95f1fa9eccdecdc3c6551b8dc4d169073cee it doesn't look likely that that's going to happen either.

TLDR you probably can't use poetry and mkl-umath together.
